examples/tutorials/nb_python/Habitat_Interactive_Tasks.py

```diff
@@ -0,0 +1,2078 @@
+# @markdown - modify the `step` function of the simulator to use the `grab/release` action we define earlier.
+
+# @markdown #### Writing the `step` function:
+# @markdown Since we added a new action for this task, we have to modify the `step` function to define what happens when `grab/release` action is called. If a simple navigation action (`move_forward`, `turn_left`, `turn_right`) is called, we pass it forward to `act` function of the agent which already defines the behavior of these actions.
+
+# @markdown For the `grab/release` action, if the agent is not already holding an object, we first call the `raycast` function using the values from the `ActuationSpec` to see if any object is grippable. If it returns a valid object id, we put the object in a "invisible" inventory and remove it from the scene.
+
+# @markdown If the agent was already holding an object, `grab/release` action will try release the object at the same relative position as it was grabbed. If the object can be placed without any collision, then the `release` action is successful.
+
+from habitat.sims.habitat_simulator.habitat_simulator import HabitatSim
+from habitat_sim.nav import NavMeshSettings
+from habitat_sim.utils.common import quat_from_coeffs, quat_to_magnum
+
+
+@registry.register_simulator(name="RearrangementSim-v0")
+class RearrangementSim(HabitatSim):
+    r"""Simulator wrapper over habitat-sim with
+    object rearrangement functionalities.
+    """
+
+    def __init__(self, config: Config) -> None:
+        self.did_reset = False
+        super().__init__(config=config)
+        self.grip_offset = np.eye(4)
+
+        agent_id = self.habitat_config.DEFAULT_AGENT_ID
+        agent_config = self._get_agent_config(agent_id)
+
+        self.navmesh_settings = NavMeshSettings()
+        self.navmesh_settings.set_defaults()
+        self.navmesh_settings.agent_radius = agent_config.RADIUS
+        self.navmesh_settings.agent_height = agent_config.HEIGHT
+
+    def reconfigure(self, config: Config) -> None:
+        super().reconfigure(config)
+        self._initialize_objects()
+
+    def reset(self):
+        sim_obs = super().reset()
+        if self._update_agents_state():
+            sim_obs = self.get_sensor_observations()
+
+        self._prev_sim_obs = sim_obs
+        self.did_reset = True
+        self.grip_offset = np.eye(4)
+        return self._sensor_suite.get_observations(sim_obs)
+
+    def _initialize_objects(self):
+        objects = self.habitat_config.objects[0]
+        obj_attr_mgr = self.get_object_template_manager()
+
+        # first remove all existing objects
+        existing_object_ids = self.get_existing_object_ids()
+
+        if len(existing_object_ids) > 0:
+            for obj_id in existing_object_ids:
+                self.remove_object(obj_id)
+
+        self.sim_object_to_objid_mapping = {}
+        self.objid_to_sim_object_mapping = {}
+
+        if objects is not None:
+            object_template = objects["object_template"]
```

Comment:
I would expect `objects` is a list, but here it's just one object. Little confusing.

Reply:
Yeah, for the purpose of this tutorial, I was just doing one object. The point was that these can easily be generalized to more complex scenarios but I wanted to keep all downstream reward computation simple.